File: indexer-go/internal/combat/dna.go
// Package combat is the Go port of the on-chain round-based battle simulator
// (contracts/ethereum/src/CombatSim.sol and the matching Rust combat.rs).
// It exists so indexer-go and the UI can replay a settled battle and estimate
// pre-fight win odds without an RPC simulation, deriving the exact same result
// the chain did from the same (dna, rarity, level, skill, seed) inputs.
//
// Every function here is pure integer math, bit-identical to both contracts —
// cross-chain parity is enforced by the golden vectors in
// contracts/test-vectors/battle.json (see combat_golden_test.go). If a vector
// fails, this port is wrong; fix the Go, never the vector.
package combat

// Attrs are the level-scaled, rarity-multiplied battle attributes derived from
// a pet's DNA (plan §3.1). Mirrors DnaLib.Attrs / dna::Attrs.
type Attrs struct {
	HP      uint16
	ATK     uint16
	DEF     uint16
	INT     uint16 // magic attack + initiative + crits
	MDEF    uint16
	Element uint8 // 0-5
}

// digitPair returns the two-digit value at pairIdx (0-indexed, LSB-first):
// (dna / 100^pairIdx) % 100.
func digitPair(dna uint64, pairIdx uint32) uint64 {
	div := uint64(1)
	for range pairIdx * 2 {
		div *= 10
	}
	return (dna / div) % 100
}

// Extract derives level-scaled, rarity-multiplied battle attributes from dna
// (plan §3.1). All intermediate math is u64 to match the contracts before the
// final u16 truncation.
func Extract(dna uint64, rarity uint8, level uint16) Attrs {
	elem := digitPair(dna, 0) % 6
	hpGene := digitPair(dna, 1)
	atkGene := digitPair(dna, 2)
	defGene := digitPair(dna, 3)
	intGene := digitPair(dna, 4)
	mdefGene := digitPair(dna, 5)

	r := max(uint64(rarity), 1)
	mul := 100 + (r-1)*5
	lv := uint64(level)

	return Attrs{
		HP:      uint16((100 + 4*hpGene + 6*lv) * mul / 100),
		ATK:     uint16((10 + atkGene + 2*lv) * mul / 100),
		DEF:     uint16((10 + defGene + 2*lv) * mul / 100),
		INT:     uint16((10 + intGene + 2*lv) * mul / 100),
		MDEF:    uint16((10 + mdefGene + 2*lv) * mul / 100),
		Element: uint8(elem),
	}
}

// elementMod is the element advantage multiplier (out of 100) for a strike
// from attacker onto defender on the six-element wheel (plan §3.2): 115
// (advantage), 85 (disadvantage), or 100 (neutral/same).
func elementMod(attacker, defender uint8) uint64 {
	if attacker == defender {
		return 100
	}
	if defender == (attacker+1)%6 {
		return 115 // attacker hits its next → advantage
	}
	if attacker == (defender+1)%6 {
		return 85 // defender is attacker's next → disadvantage
	}
	return 100 // non-adjacent in the 6-cycle → neutral
}

// resolveSpecies resolves a pet's species id from its DNA cosmetic digit-pair
// and its rarity tier's pool size (plan §3.7), mirroring dna::resolve_species /
// PetCore._resolveSpecies. poolSizes is indexed by rarity-1 (clamped); a pool
// size of 0 means "no species" (id 0).
func resolveSpecies(dna uint64, rarity uint8, poolSizes [5]uint8) uint16 {
	idx := 0
	if rarity > 1 {
		idx = int(rarity - 1)
	}
	if idx > 4 {
		idx = 4
	}
	poolSize := poolSizes[idx]
	if poolSize == 0 {
		return 0
	}
	return uint16(digitPair(dna, 6) % uint64(poolSize))
}

File: protocol/src/combat/dna.ts
/**
 * TypeScript port of the fight-relevant half of indexer-go's combat package
 * (dna.go), itself a port of contracts/ethereum/src/DnaLib.sol / combat.rs
 * (Solana). Only what CombatSim.simulate needs is ported here — species
 * resolution (dna.go's resolveSpecies) happens at mint time in PetCore, not
 * inside the battle sim, so it's out of scope for this port.
 *
 * Every function here is pure bigint math, bit-identical to the Solidity /
 * Rust / Go implementations — cross-chain (and now cross-runtime) parity is
 * enforced by the golden vectors in contracts/test-vectors/battle.json (see
 * protocol/tests/combat/goldenVectors.test.ts). If a vector fails, this
 * port is wrong; fix the TS, never the vector.
 */

/** Level-scaled, rarity-multiplied battle attributes derived from a pet's DNA
 *  (plan §3.1). Mirrors DnaLib.Attrs (Solidity) / dna::Attrs (Rust) /
 *  indexer-go's Attrs (Go). All fields are bigint: DNA is a 16-digit number
 *  that exceeds Number.MAX_SAFE_INTEGER, so this module never uses `number`
 *  for anything that participates in the sim's arithmetic. */
export interface Attrs {
    hp: bigint;
    atk: bigint;
    def: bigint;
    int: bigint; // magic attack + initiative + crits
    mdef: bigint;
    element: bigint; // 0-5
}

const UINT16_MOD = 65536n;

/** Truncates to the low 16 bits, matching Solidity's `uint16(...)` cast /
 *  Go's `uint16(...)` conversion (wraparound, not a clamp). Never actually
 *  wraps for realistic pet stats (level <= 100, rarity 1-5) — kept for exact
 *  parity with the reference implementations regardless. */
export function toUint16(x: bigint): bigint {
    return x % UINT16_MOD;
}

/** digitPair returns the two-digit value at pairIdx (0-indexed, LSB-first):
 *  (dna / 100^pairIdx) % 100. */
export function digitPair(dna: bigint, pairIdx: number): bigint {
    const div = 100n ** BigInt(pairIdx);
    return (dna / div) % 100n;
}

/** Derives level-scaled, rarity-multiplied battle attributes from dna (plan
 *  §3.1). Mirrors DnaLib.extract / dna::extract / indexer-go's Extract. */
export function extract(dna: bigint, rarity: number, level: number): Attrs {
    const elem = digitPair(dna, 0) % 6n;
    const hpGene = digitPair(dna, 1);
    const atkGene = digitPair(dna, 2);
    const defGene = digitPair(dna, 3);
    const intGene = digitPair(dna, 4);
    const mdefGene = digitPair(dna, 5);

    const r = BigInt(Math.max(rarity, 1));
    const mul = 100n + (r - 1n) * 5n;
    const lv = BigInt(level);

    return {
        hp: toUint16(((100n + 4n * hpGene + 6n * lv) * mul) / 100n),
        atk: toUint16(((10n + atkGene + 2n * lv) * mul) / 100n),
        def: toUint16(((10n + defGene + 2n * lv) * mul) / 100n),
        int: toUint16(((10n + intGene + 2n * lv) * mul) / 100n),
        mdef: toUint16(((10n + mdefGene + 2n * lv) * mul) / 100n),
        element: elem,
    };
}

/** Element-advantage multiplier (out of 100) for a strike from attacker onto
 *  defender on the six-element wheel (plan §3.2): 115 (advantage), 85
 *  (disadvantage), or 100 (neutral/same). */
export function elementMod(attacker: bigint, defender: bigint): bigint {
    if (attacker === defender) return 100n;
    if (defender === (attacker + 1n) % 6n) return 115n; // attacker hits its next -> advantage
    if (attacker === (defender + 1n) % 6n) return 85n; // defender is attacker's next -> disadvantage
    return 100n; // non-adjacent in the 6-cycle -> neutral
}
